aten/src/ATen/native/sparse/SparseFactories.cpp

```diff
@@ -0,0 +1,176 @@
+
+namespace {
+SparseTensor _spdiags_sparse_cpu_coo(
+    const Tensor& diagonals,
+    const Tensor& offsets,
+    IntArrayRef shape) {
+  TensorOptions sparse_options = diagonals.options().layout(kSparse);
+  TensorOptions indices_options =
+      offsets.options().device(sparse_options.device());
+
+  const int64_t n_rows_out = shape[0];
+  const int64_t n_cols_out = shape[1];
+  const int64_t n_cols_in = diagonals.size(1);
+  const int64_t n_diag = diagonals.size(0);
+
+  // This is the the largest number of values we could set along a diagonal, it
+  // is the length of an input row or the length of the longest diagonal,
+  // whichever is smaller
+  const int64_t min_PQN = std::min(std::min(n_rows_out, n_cols_out), n_cols_in);
+  // Conservative estimate, if all diagonals we are assigning to are of the max
+  // size
+  const int64_t max_nnz = n_diag * min_PQN;
+  // Note: when offsets are positive we push the start point in the row of D
+  // forward.
+  Tensor indices = at::empty({2, max_nnz}, indices_options);
+  Tensor values = at::empty({max_nnz}, diagonals.options());
+
+  // Get accessors on input, and output
+  auto indices_accessor = indices.accessor<int64_t, 2>();
+  auto offsets_accessor = offsets.accessor<int64_t, 1>();
+  int64_t actual_nnz = 0;
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(
+      at::ScalarType::BFloat16,
+      at::ScalarType::Half,
+      at::ScalarType::Bool,
+      at::ScalarType::ComplexHalf,
+      values.scalar_type(),
+      "spdiags",
+      [&] {
+        auto values_accessor = values.accessor<scalar_t, 1>();
+        auto const diagonals_accessor = diagonals.accessor<scalar_t, 2>();
+
+        for (const auto d_i : c10::irange(n_diag)) {
+          const int64_t& off_i = offsets_accessor[d_i];
+          const int64_t row_out_begin = off_i < 0 ? std::abs(off_i) : 0;
+          const int64_t col_out_begin = off_i > 0 ? off_i : 0;
+
+          // Number of column, row positions we can assign into and the number
+          // of diag positions we can read from
+          const int64_t row_slots = n_rows_out - row_out_begin;
+          const int64_t col_slots = n_cols_out - col_out_begin;
+          const int64_t diag_slots = n_cols_in - col_out_begin;
+
+          // Max number of reads lets us limit the loop by the first thing we
+          // will exhaust, note if any of these comes up negative irange will be
+          // empty so we don't have to handle that case explicitly
+          const int64_t max_read =
+              std::min(diag_slots, std::min(row_slots, col_slots));
+
+
+          for (const auto read_count : c10::irange(max_read)) {
+            const auto col = col_out_begin + read_count;
+            const auto row = row_out_begin + read_count;
+
+            values_accessor[actual_nnz] = diagonals_accessor[d_i][col];
+            indices_accessor[0][actual_nnz] = row;
+            indices_accessor[1][actual_nnz] = col;
+            ++actual_nnz;
+          }
+        }
+      });
```

Comment:
We will need these `actual_nnz` per diagonal for the CUDA version anyway, just to make sure we know the result shape to be able to do a single device sync, and to avoid any atomic operations.